CMakeLists.txt

```diff
@@ -0,0 +1,15 @@
+project(miopen-mlir)
+cmake_minimum_required(VERSION 3.13.4)
+
+set(LLVM_ENABLE_PROJECTS "llvm;mlir;lld" CACHE STRING "List of default llvm targets")
+set(LLVM_BUILD_EXAMPLES ON CACHE BOOL "")
+set(LLVM_TARGETS_TO_BUILD "X86;NVPTX;AMDGPU" CACHE STRING "")
+set(LLVM_ENABLE_ASSERTIONS ON CACHE BOOL "")
+set(BUILD_SHARED_LIBS ON CACHE BOOL "")
+set(LLVM_BUILD_LLVM_DYLIB ON CACHE BOOL "")
+set(MLIR_ROCM_RUNNER_ENABLED ON CACHE BOOL "")
+set(MLIR_MIOPEN_DRIVER_ENABLED ON CACHE BOOL "")
+set(MLIR_ENABLE_SQLITE ON CACHE BOOL "")
```

Comment:
This will be generally used for build and test. I guess we should have a build config for library export.

Reply:
cget does 'cmake ..' which builds according to the settings in the top-level CMakeLists.txt.

These added 'install' lines ensure the libMLIRMIOpenThin.so and Miir.h get installed in the specified install path (for MIGraph it is /usr/local).

Reply:
I don't see libMLIRMIOpen generated in my build or exported.. Is that because I have BUILD_SHARED_LIBS=ON?

Reply:
That's right. In order for that standalone lib to be exported, you will need all 3 below flags:

  -DBUILD_SHARED_LIBS=OFF \
  -DLLVM_BUILD_LLVM_DYLIB=OFF \
  -DLLVM_ENABLE_TERMINFO=OFF

Reply:
Indeed it is.. Just updated with your changes.

Reply:
Sounds good. Just a reminder, I think you'd still need to drop a few less relevant flags, because the custom build static library target is going to aggregate all static libraries together. We need to minimize the number of targets built. I think below will do the work

```cmake
set(LLVM_ENABLE_PROJECTS "mlir;lld" CACHE STRING "List of default llvm targets")
set(BUILD_SHARED_LIBS OFF CACHE BOOL "")
set(LLVM_BUILD_LLVM_DYLIB OFF CACHE BOOL "")
set(LLVM_ENABLE_TERMINFO OFF CACHE BOOL "")
set(MLIR_ENABLE_SQLITE ON CACHE BOOL "")
```

I used to play around with the list of flags got turned on here and find out that you can easily make the static lib 2x the size with turning on some irrelevant ones.

Reply:
I assume this should also be set minimally:

> set(LLVM_TARGETS_TO_BUILD "AMDGPU" CACHE STRING "")

Do we need X86 also for host codegen? The library wouldn't need it but the driver does.

Reply:
I might be wrong but I suppose AMDGPU is for building the backend. We don't need to backend to run those kernels in the library. Same with X86, the library won't need the driver so we don't need to build it too.

Reply:
Both seems to be needed, the drivers are built unconditionally.

Reply:
We should also have a bundled libMLIRMIOpen.so to share across invocations.

Reply:
> Both seems to be needed, the drivers are built unconditionally.

In the latest of `miopen-dialect` I am able to do the following without a problem:
```bash
cmake -G "Unix Makefiles" ../llvm \
      -DLLVM_ENABLE_PROJECTS="mlir;lld" \
      -DCMAKE_BUILD_TYPE=Release \
      -DBUILD_SHARED_LIBS=OFF \
      -DLLVM_BUILD_LLVM_DYLIB=OFF \
      -DLLVM_ENABLE_TERMINFO=OFF
make -j$(nproc) libMLIRMIOpen
```
In the build result, I can find the `libMLIRMIOpen.a` and cannot find `mlir-miopen-driver`.

> We should also have a bundled libMLIRMIOpen.so to share across invocations.

You mean make shared library and allow dynamic loading? Is it required by MIgraphx?

Reply:
I think that would build all backends.. I see WebAssembly,XCore,SPARC, etc. getting built.

Reply:
@sjw36 With the trimmed flags, the library I build is 270 Mb and is able to functional just fine. Did you get the 9 Gb library with the same set of flags?

Reply:
Hmm.. I must be building Debug..

Reply:
Yep, release is much smaller. Still has many external LLVM symbols (which would be DCE'd) and does incorporate all backends. With LLVM_TARGETS_TO_BUILD=X86;AMDGPU, the library size is ~200Mb.

Reply:
Thanks for confirming!